### DISCUSSION TITLE
Trying to use prisma with pnpm

### DISCUSSION THREAD

**Discussion by hqnna** [prisma/prisma#12501] [Q&A] [answered]
My [current project](../../../xpyxel/codebin) uses PNPM for package management, when trying to run `prisma generate` which seems to use NPM no matter which package manager is used, it breaks and is unable to actually generate the client, is there any way to get around this or am i SOL?

![image](https://user-images.githubusercontent.com/46457601/159834402-aeeea8a1-44f3-4714-88c2-08cf8d72616a.png)

**Comment by austincrim** [comment, marked as answer]
Hey @xpyxel 👋,

Do you have the Prisma CLI installed globally? I noticed you ran the command without `npx` or `pnpx`. I was able to use `pnpm` with a local Prisma project without issue using `pnpx`.

**Reply by hqnna** [reply]
I have it installed in the project locally with `pnpm i -D prisma`. But in that screenshot it was installed globally yeah, installing it locally doesn't change anything however from what i've tested.

**Reply by austincrim** [reply]
Have you tried running with `pnpx prisma generate` while installed in the local project?

**Reply by hqnna** [reply]
Even when using pnpm it seems to still use npm for whatever reason, the progress bar here is from npm:

![image](https://user-images.githubusercontent.com/46457601/160031757-a29be848-c541-423e-b015-86152de90e08.png)

**Reply by hqnna** [reply]
This can cause issues if it creates a npm lockfile and messes with the `node_modules` pnpm creates.

**Reply by janpio** [reply]
That might be a bug in our installation logic. There should be an issue for it already somewhere...
You can probably avoid this by manually installing `@prisma/client` into your project before running the command.

**Reply by hqnna** [reply]
Let me try that real quick and get back to you.

**Reply by hqnna** [reply]
Yeah that seems to bypass the npm installation step in it, thanks for the help.

**Reply by janpio** [reply]
I now also found the issue this was reported in and will refine it a bit to make it easier to find: https://github.com/prisma/prisma/issues/9848

**Reply by vicentematus** [reply]
@heyhanna did you resolve this problem? prisma migrate is really slow with `pnpm`. I have Prisma installed in the project locally.

**Reply by vicentematus** [reply]
> @heyhanna did you resolve this problem? prisma migrate is really slow with `pnpm`. I have Prisma installed in the project locally.

Okay so after i run `pnpm install @prisma/client` and then execute it with `pnpm exec prisma migrate `pnpm exec prisma migrate dev --name init` its working.

**Comment by hqnna** [comment]
Yes, manually installing client with pnpm fixes it.

**Reply by Zenthae** [reply]
same for me, but it should be noted somewhere on the doc, i had to search for some time before finding this place

**Reply by aestheticsdata** [reply]
after manually `pnpm install @prisma/client`, do we need the prisma package ?

**Reply by janpio** [reply]
Yes - you use the CLI to re-generate the Client (which uses the data from `@prisma/client`).

**Comment by ErlanBelekov** [comment]
I have a Dockerfile, and it doesn't help me to add prisma before `pnpm install`

```
FROM node:16-alpine as setup

WORKDIR /app

# install pnpm
RUN npm install -g pnpm

COPY .npmrc package.json pnpm-lock.yaml pnpm-workspace.yaml tsconfig.base.json ./

COPY packages/eslint-config/package.json ./packages/eslint-config/
COPY packages/prettier-config/package.json ./packages/prettier-config/
COPY packages/web/package.json ./packages/web/
COPY packages/shared-utils/package.json ./packages/shared-utils/

# copy prisma schema
COPY packages/server/prisma/schema.prisma ./packages/server/prisma/

COPY packages/shared-utils ./packages/shared-utils/

RUN cd packages/shared-utils && pnpm add @prisma/client@3.15.2 prisma@3.15.2

RUN pnpm install --frozen-lockfile --ignore-scripts

# run custom generate script(pnpm prisma generate)
RUN pnpm --filter "@hub/shared-utils" generate
```

this is `package.json` of `shared-utils`

```
{
  "name": "@hub/shared-utils",
  "version": "1.0.0",
  "main": "dist/src/index.js",
  "types": "dist/src/index.d.ts",
  "private": true,
  "license": "MIT",
  "scripts": {
    "build": "rm -rf dist && tsc --build",
    "test": "jest",
    "precommit": "eslint --ext .ts \"*/**/*\" --fix && prettier --write .",
    "generate": "pnpm prisma generate",
    "postinstall": "pnpm prisma generate"
  },
  "dependencies": {
    "currency.js": "^2.0.4",
    "@prisma/client": "3.15.2",
    "date-fns": "^2.28.0"
  },
  "devDependencies": {
    "@hub/eslint-config": "1.0.0",
    "@hub/prettier-config": "1.0.0",
    "prisma": "3.15.2"
  },
  "prisma": {
    "schema": "../server/prisma/schema.prisma"
  },
  "prettier": "@hub/prettier-config",
}

```

**Reply by austincrim** [reply]
What error are you getting?

**Reply by ErlanBelekov** [reply]
this is the error, I have created a separate issue for it(https://github.com/prisma/prisma/issues/14944)
```
#0 2.239
#0 2.239 > @hub/shared-utils@1.0.0 generate /app/packages/shared-utils
#0 2.239 > pnpm prisma generate
#0 2.239
#0 36.80 Prisma schema loaded from ../server/prisma/schema.prisma
#0 451.1 npm ERR! Cannot read properties of null (reading 'matches')
#0 451.2
#0 451.2 npm ERR! A complete log of this run can be found in:
#0 451.2 npm ERR!     /root/.npm/_logs/2022-08-23T02_34_57_814Z-debug-0.log
#0 451.2 Error: Command failed with exit code 1: npm install -D prisma@3.15.2
#0 451.3 undefined
#0 451.3 /app/packages/shared-utils:
#0 451.3  ERR_PNPM_RECURSIVE_RUN_FIRST_FAIL  @hub/shared-utils@1.0.0 generate: `pnpm prisma generate`
#0 451.3 Exit status 1
```

**Reply by tal-rofe** [reply]
Same for me, exact flow

**Reply by tal-rofe** [reply]
@ErlanBelekov 

Do you a have a workaround?

**Reply by austincrim** [reply]
There is a workaround in the issue linked above:

> adding public-hoist-pattern[]=*prisma* to .npmrc file fixed this issue, took the solution from [here](https://github.com/prisma/prisma/issues/12921)

**Comment by saberhosneydev** [comment]
After hours of trial and error😄 I think I found a workaround for this problem.
Steps to get everything working in pnpm:
1. install everything normally with `pnpm install`
2. adjust the generator in `prisma.schema` to the following
    ```prisma
    generator client {
        provider = "prisma-client-js"
        output   = "./prisma-client" //the value to adjust or add if you dont have it
    }
    ```
3. Run the generators with `pnpm exec prisma generate` it will create the prisma client in the output folder we specified which should be `<project_root>/prisma/prisma-client`
4. Now the **important** part run `pnpm add ./prisma/prisma-client` this will make a link to the generated client.

now you can use it with `import { <whatYouWant> } from '.prisma/client'` 
I know this solution is not optimal since if you have an existing project you will have to edit the imports in your files, I still believe its a good solution since you can automate this with a script or integrate it with your own automation tool.

*`bonus tip:`*
add prisma-client folder to `.gitignore` so you don't flood your VCS with non-needed files ( it will get automatically generated when you do the 3rd step )
```diff
...
+prisma/prisma-client
```

**Reply by meera** [reply]
Worked like a charm!

**Reply by yuanguandong** [reply]
👍🏻

**Reply by rohanrajpal** [reply]
we have to add `pnpm add` in the dockerfile too?

**Reply by nocategory** [reply]
holy moly, what a precious comment that fixed the issues I was having for weeks 🤣

**Reply by BryanGT** [reply]
thanks! I spend like 2 days trying in too many ways to fixed, and finally this works! God bless you 💯

**Reply by artemis-prime** [reply]
Didn't work for me

**Reply by Lukas-Blom** [reply]
LETS FUCKING GO! Thank you for sharing this information!!

**Comment by anatoliy-t7** [comment]
`pnpm add @prisma/client` fixed for me

**Reply by MIREHMAN** [reply]
where to add it ??

**Reply by ludralph** [reply]
The command will be issued on the terminal or command line interface. Once it completes execution,  `@prisma/client` will be added to the dependencies section of package.json file

**Comment by ExStu** [comment]
`dependencies:
@prisma/client 6.1.0

devDependencies:
prisma 6.1.0`

pnpm
`9.15.0`

node
`v22.11.0`

this had fixed the issue for me

`generator client {
  provider = "prisma-client-js"
  output   = "../node_modules/.prisma/client"
}`

**Reply by lynx-r** [reply]
This answer helped me.

**Comment by ryicoh** [comment]
prisma: "6.2.1"
pnpm: 9.15.4
node: v22.13.1

```schema.prisma
generator client {
  provider = "prisma-client-js"
  output   = "../../node_modules/.pnpm/.prisma/client"
}
```

```json:package.json
"prisma:generate": "prisma generate && pnpm prisma:link",
"prisma:link": "ln -sf ../../.prisma ../node_modules/.pnpm/@prisma+client@6.2.1_prisma@6.2.1/node_modules/.prisma",
```

it works for me
